Sources/IRGen/Function/FunctionContext.swift

```diff
@@ -21,31 +22,92 @@ class FunctionContext {
   /// Whether the function is declared in a struct.
   var isInStructFunction: Bool
 
+  /// Stack of do-catch statements, for handling errors.
   var doCatchStatementStack: [DoCatchStatement]
 
+  /// Stack of code blocks ({ ... })
+  var blockStack: [YUL.Block]
+
+  /// Fresh variable counter.
+  private var counter: Int
+
   init(environment: Environment,
        scopeContext: ScopeContext,
        enclosingTypeName: String,
-       isInStructFunction: Bool,
-       doCatchStatementStack: [DoCatchStatement] = []) {
+       isInStructFunction: Bool) {
     self.environment = environment
     self.scopeContext = scopeContext
     self.enclosingTypeName = enclosingTypeName
     self.isInStructFunction = isInStructFunction
-    self.doCatchStatementStack = doCatchStatementStack
+
+    self.doCatchStatementStack = []
+    self.blockStack = [Block([])]
+    self.counter = 0
   }
 
-  func push(doCatch stmt: DoCatchStatement) {
-    doCatchStatementStack.append(stmt)
+  func emit(_ statement: YUL.Statement) {
+    let catchableSuccesses = statement.catchableSuccesses
+    if catchableSuccesses.count > 0 {
+      let allSucceeded = catchableSuccesses.reduce(.literal(.num(1)), { acc, success in
+        .functionCall(FunctionCall("and", [acc, success]))
+      })
+      let allSucceededVariable = freshVariable()
+      emit(.inline("let \(allSucceededVariable) := \(allSucceeded.description)"))
+      emit(.inline("switch \(allSucceededVariable)"))
+      emit(.inline("case 0"))
+      emit(.block(withNewBlock {
+        topDoCatch!.catchBody.forEach { statement in
+          emit(IRStatement(statement: statement).rendered(functionContext: self))
+        }
+      }))
+      // Further statements will be inside this block. It will be closed
+      // eventually in withNewBlock(...).
+      emit(.inline("case 1"))
+      _ = pushBlock()
+    }
+    blockStack[blockStack.count - 1].statements.append(statement)
   }
 
-  @discardableResult
-  func pop() -> DoCatchStatement? {
-    return doCatchStatementStack.popLast()
+  func withNewBlock(_ inner: () -> Void) -> YUL.Block {
+    // The inner() call may cause more blocks to be pushed (see emit above),
+    // so here we make sure to pop and emit all the additional blocks before
+    // returning this one.
+    let outerCount = pushBlock()
+    inner()
+    while blockStack.count != outerCount {
+      emit(.block(popBlock()))
+    }
+    return popBlock()
   }
 
-  var top: DoCatchStatement? {
-    return doCatchStatementStack.last
+  func pushBlock() -> Int {
+    blockStack.append(Block([]))
+    return blockStack.count
+  }
+
+  func popBlock() -> YUL.Block {
+    return self.blockStack.popLast()!
+  }
+
+  func freshVariable() -> String {
+    let varName = "$temp\(self.counter)"
+    self.counter += 1
+    return varName
   }
 
+  func dump() -> String {
```

Comment:
Do we need this? Could you comment what this does if you intend to keep this function?